agent/engine/task_manager_test.go

```diff
@@ -15,19 +15,436 @@
 package engine
 
 import (
+	"sync"
 	"testing"
 	"time"
 
 	"github.com/aws/amazon-ecs-agent/agent/api"
 	"github.com/aws/amazon-ecs-agent/agent/engine/dockerstate/mocks"
 	"github.com/aws/amazon-ecs-agent/agent/engine/testdata"
+	"github.com/aws/amazon-ecs-agent/agent/eventstream"
 	"github.com/aws/amazon-ecs-agent/agent/statemanager"
 	"github.com/aws/amazon-ecs-agent/agent/utils/ttime/mocks"
 	"github.com/stretchr/testify/assert"
 
 	"github.com/golang/mock/gomock"
+	"golang.org/x/net/context"
 )
 
+func TestContainerNextStateFromNone(t *testing.T) {
```

Comment:
I think it'd actually be good to rewrite all of these into [table-driven](https://github.com/golang/go/wiki/TableDrivenTests) [subtests](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks).

```go
var statetests = []struct {
	containerCurrentStatus api.ContainerStatus
	containerDesiredStatus api.ContainerStatus
	expectedContainerStatus api.ContainerStatus
	expectedTransitionActionable bool
	expectedTransitionPossible bool
}{
	// NONE -> RUNNING transition is allowed and actionable, when desired is Running.
	// The expected next status is PULLED
	{api.ContainerStatusNone, api.ContainerRunning, api.ContainerPulled, true, true},

	// ...
}

func TestStateChanges(t *testing.T) {
	for _, statetest := range statetests {
		t.Run(fmt.Sprintf("%s->%s", statetest.containerCurrentStatus, containerDesiredStatus),
			func (t *testing) {
				testContainerNextStateAssertions(t,
					containerCurrentStatus,
					containerDesiredStatus,
					expectedContainerStatus,
					expectedTransitionActionable,
					expectedTransitionPossible)
			}
		)
	}
}
```